pkg/curatedpackages/bundle.go

```diff
@@ -0,0 +1,117 @@
+package curatedpackages
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+
+	api "github.com/aws/eks-anywhere-packages/api/v1alpha1"
+	"github.com/aws/eks-anywhere-packages/pkg/artifacts"
+	"github.com/aws/eks-anywhere-packages/pkg/bundle"
+	"github.com/aws/eks-anywhere-packages/pkg/testutil"
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/executables"
+)
+
+const (
+	Cluster         = "cluster"
+	Registry        = "registry"
+	RegistryBaseRef = "public.ecr.aws/q0f6t3x4/eksa-package-bundles"
+)
+
+func GetLatestBundle(ctx context.Context, kubeConfig string, source string, kubeVersion string) (*api.PackageBundle, error) {
+	var (
+		packageBundle *api.PackageBundle
+		err           error
+	)
+
+	switch strings.ToLower(source) {
+	case Cluster:
+		packageBundle, err = getActiveBundleFromCluster(ctx, kubeConfig)
+	case Registry:
+		packageBundle, err = getLatestBundleFromRegistry(ctx, kubeVersion)
+	}
```

Comment:
I wouldn't panic. We don't know at this point what supplied the incorrect source so we should still return an error. 

To appreciate why we can look at what actually happens: the user supplies the source. That's a user input error so I'd expect the error to bubble up and exit the program gracefully. The sticking point though is that this function has insufficient context to justify a panic. 

```go
return nil, errors.New("unknown source")
```

Reply:
Chris is totally right, if this were to stay a string variable, then that would be the case. If source becomes a user type, as I think it should, then I'd still panic because that validation of user input should have already happened, and there's really no reason there should ever be an unexpected value here, and if there is, something is very wrong. I'm a fan of fail early and loudly. Some folks aren't. 

I'm 100% fine with returning an error though. So long as it's identified and not swallowed, I'm happy.

Reply:
Hmm... If you move to a user type, then you could get rid of this entirely...

(Pseudo-code follows)

```go
type bundleGetter interface {
    getLatestBundle(ctx context.Context, kubeConfig, kubeVersion string) (*Bundle, err)
}
func (c Cluster) getLatestBundle(ctx context.Context, kubeConfig, kubeVersion string) (*Bundle, err) {
    // implement me
}
func (r Registry) getLatestBundle(ctx context.Context, kubeConfig, kubeVersion string) (*Bundle, err) {
    // implement me
}
```

With the above, then this switch goes away and becomes:

```go
bundle, err := source.getLatestBundle(ctx, kubeConfig, kubeVersion)
```

https://lostechies.com/derickbailey/2008/11/26/ptom-descriptive-state-enumeration/